File: macos/TokendialCore/Sources/TokendialCore/Install/InstallCatalog.swift
import Foundation

/// The install recipes, read from the provider specs the app carries. A spec without an install block, like
/// GLM, has no recipe and its row simply never offers to install anything.
///
/// The directory is handed in rather than found, exactly as `Strings` takes its catalogue: the app passes the
/// bundle's copy of `docs/providers`, the tests pass the one in the repository.
public enum InstallCatalog {
    public private(set) static var all: [String: InstallRecipe] = [:]

    public static func load(from directory: URL) {
        let files = (try? FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil)) ?? []
        var found: [String: InstallRecipe] = [:]
        for file in files where file.pathExtension == "json" && file.lastPathComponent != "schema.json" {
            guard let data = try? Data(contentsOf: file), let root = JSON.object(data), let id = root.str("id"),
                  let recipe = parse(id, root) else { continue }
            found[id] = recipe
        }
        all = found
    }

    /// A profile shares the recipe of the tool it is an account of: claude-work installs Claude Code.
    public static func recipe(_ providerId: String) -> InstallRecipe? { all[ProviderFamily.of(providerId)] }

    private static func parse(_ id: String, _ root: JSONObject) -> InstallRecipe? {
        guard let install = root.obj("install"),
              let vendor = install.str("vendor"),
              let docsUrl = install.str("docsUrl"),
              let mac = install.obj("macos"),
              let kind = InstallKind(rawValue: mac.str("kind") ?? ""),
              let line = mac.str("install") else { return nil }
        let detect = mac.obj("detect") ?? [:]
        var step: SignInStep?
        if let signIn = mac.obj("signIn"), let command = signIn.str("command"), let hint = signIn.str("hint") {
            step = SignInStep(command: command, hint: hint)
        }
        return InstallRecipe(
            providerId: id,
            vendor: vendor,
            docsUrl: docsUrl,
            kind: kind,
            detect: Detect(commands: mac.obj("detect")?.arr("commands").compactMap { $0 as? String } ?? [],
                           paths: detect.arr("paths").compactMap { $0 as? String }),
            requires: mac.arr("requires").compactMap { $0 as? String },
            install: line,
            signIn: step,
            downloadUrl: mac.str("downloadUrl"))
    }
}

File: windows/Tokendial.Core/Install/InstallCatalog.cs
using System.Reflection;
using System.Text.Json;
using Tokendial.Core.Providers;

namespace Tokendial.Core.Install;

/// <summary>The install recipes read from the provider specs embedded in this assembly. A spec without an install block, like GLM, has no recipe.</summary>
public static class InstallCatalog
{
    private static readonly Lazy<IReadOnlyDictionary<string, InstallRecipe>> all = new(Load);

    public static IReadOnlyDictionary<string, InstallRecipe> All => all.Value;

    /// <summary>The recipe for a provider; Claude profiles (claude-&lt;slug&gt;) share Claude's.</summary>
    public static InstallRecipe? For(string providerId) => All.GetValueOrDefault(Family(providerId));

    public static string Family(string providerId) => ProviderFamily.Of(providerId);

    private static IReadOnlyDictionary<string, InstallRecipe> Load()
    {
        var assembly = typeof(InstallCatalog).Assembly;
        var recipes = new Dictionary<string, InstallRecipe>(StringComparer.Ordinal);
        foreach (var name in assembly.GetManifestResourceNames().Where(n => n.StartsWith("providers/", StringComparison.Ordinal) && n != "providers/schema.json"))
        {
            using var stream = assembly.GetManifestResourceStream(name)!;
            using var document = JsonDocument.Parse(stream);
            var root = document.RootElement;
            if (!root.TryGetProperty("install", out var install)) continue;
            var id = root.GetProperty("id").GetString()!;
            recipes[id] = new InstallRecipe(id, install.GetProperty("vendor").GetString()!, install.GetProperty("docsUrl").GetString()!,
                Platform(install.GetProperty("windows")), Platform(install.GetProperty("macos")));
        }
        return recipes;
    }

    private static PlatformRecipe Platform(JsonElement element)
    {
        var detect = element.TryGetProperty("detect", out var d) ? new Detect(Texts(d, "commands"), Texts(d, "paths")) : Detect.None;
        var signIn = element.TryGetProperty("signIn", out var s) ? new SignInStep(s.GetProperty("command").GetString()!, s.GetProperty("hint").GetString()!) : null;
        var kind = element.GetProperty("kind").GetString() == "app" ? InstallKind.App : InstallKind.Cli;
        var download = element.TryGetProperty("downloadUrl", out var u) ? u.GetString() : null;
        return new PlatformRecipe(kind, detect, Texts(element, "requires"), element.GetProperty("install").GetString()!, signIn, download);
    }

    private static IReadOnlyList<string> Texts(JsonElement element, string name) =>
        element.TryGetProperty(name, out var array) && array.ValueKind == JsonValueKind.Array
            ? array.EnumerateArray().Select(x => x.GetString() ?? "").Where(x => x.Length > 0).ToList()
            : [];
}
